clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1797,7 +1842,22 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     Expr *ILE = new (SemaRef.getASTContext())
         InitListExpr(SemaRef.getASTContext(), SourceLocation(), ArrayInitExprs,
                      SourceLocation());
-    ILE->setType(FD->getType());
+
+    // We need to find the type of the element for which we are generating the
+    // InitListExpr. For example, for a multi-dimensional array say a[2][3][2],
+    // the types for InitListExpr of the array and its 'sub-arrays' are -
+    // int [2][3][2], int [3][2] and int [2]. This loop is used to obtain this
```

Comment:
Shouldn't the second type be int [2][3]?   How is it [3][2]?

Reply:
I don't think so, see: https://godbolt.org/z/enqsf3

The init list exprs look like:

      `-VarDecl <col:5, col:87> col:9 arr 'int [2][3][4]' cinit
        `-InitListExpr <col:24, col:87> 'int [2][3][4]'
          |-InitListExpr <col:25, col:49> 'int [3][4]'
          | |-InitListExpr <col:25, col:31> 'int [4]


Recall that int[2][3][4] means "an array of 2 arrays of 3 of 4 ints".

Reply:
Thanks for the godbolt example.   That helped.

Reply:
It should be [3][2]. I think  int a[2][3][2] = { { {1,2},{3,4},{5,6} } , { {7,8},{9,10},{11,12} } }

Reply:
STAHP using a[2][3][2] as an example :)